### PR TITLE
refactor(BUX-611) codebase improvements

### PR DESCRIPTION
**Pull request by chris-4chain**
In this PR, breaking changes have been introduced, affecting client initialization, code organization, logging, and error handling. Here's what has been done:

- The ClientOptions type has been refined and simplified to only allow keys as strings.
   - Using TypeScript, the options are validated. For example, if a user provides xPriv with signRequest: false, a TypeScript error will be thrown.
   - Descriptive comments have been added to the fields.
   - During client creation, the logger will report which key is being used, depending on the log level.
   - In cases of incorrect configuration, the client constructor will throw an exception.
- The code responsible for HTTP requests has been relocated to a Requester class.
- Custom exception classes have been created to streamline error catching and make them distinguishable to users.
- A new Prettier configuration has been applied.
- All code segments previously marked with ts-ignore have been revised and corrected.
- The handling of `serverUrl` has been delegated to the Requester, meaning that when using http.Request, there is no need to prepend it.

Due to the extent of the changes, I believe it's better to review the current state of the code locally, as the comparison between the old and new code might not be clear enough.

# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet-js-client/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet-js-client/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [x] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -786,45 +748,39 @@ export class SpvWalletClient {
...
-      this.logger.error(Err.message)
-      throw Err
+    if (!this?.xPriv) {
+      throw new ErrorNoXPrivToSignTransaction();
     }
 
-    const Input = bsv.Transaction.Input
-    const xPriv = this.options.xPriv as bsv.HDPrivateKey
-    const txDraft = new bsv.Transaction(draftTransaction.hex)
+    const Input = bsv.Transaction.Input;
```
Comment (dorzepowski) [inline src/client.ts:755]: ### minor
```suggestion
    const input = bsv.Transaction.Input;
```

File: src/client.ts [hunk 2]
```diff
@@ -36,98 +44,52 @@ import { Logger, LoggerConfig, makeLogger, defaultLogger } from './logger/logger
...
+    if (!options.signRequest && options.xPub) {
+      this.logger.info('Using XPub. Admin requests and SendToRecipients function will not be available.');
+      return Requester.CreateXPubRequester(this.logger, serverUrl, options.xPub);
+    }
 
-    if (options.adminKey) {
-      this.SetAdminKey(options.adminKey)
+    if (!options.signRequest) {
+      throw new ErrorWithDisabledSignRequest(this.logger, options);
     }
```
Comment (dorzepowski) [inline src/client.ts:69]: Actually if you choose disabled signRequests, you still could use xPriv as key and then we can calculate xpub from it.

File: src/client.ts [hunk 3]
```diff
@@ -36,98 +44,52 @@ import { Logger, LoggerConfig, makeLogger, defaultLogger } from './logger/logger
...
-      options.xPub = bsv.HDPublicKey.fromString(options.xPubString)
-    } else if (options.accessKey) {
-      options.xPriv = undefined
-      options.xPrivString = undefined
-      options.xPub = undefined
-      options.xPubString = undefined
-      options.accessKeyString = options.accessKey.toString()
+    this.logger = makeLogger(loggerConfig);
+
+    if (options.signRequest && options.xPriv) {
```
Comment (dorzepowski) [inline src/client.ts:55]: Why this `if` statement is outside the initRequester?

File: src/requester.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
+import bsv from 'bsv';
+import { AuthHeader, setSignature } from './authentication';
+import { Logger } from './logger';
+import { ErrorNoAdminKey, ErrorResponse } from './errors';
+
+export class Requester {
```
Comment (dorzepowski) [inline src/requester.ts:6]: ### minor
maybe `HttpClient` or `ApiClient` or leave it if you don't like my propositions 🙂

File: src/requester.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  private logger: Logger;
+
+  private adminKey?: bsv.HDPrivateKey;
+  private signingKey?: bsv.HDPrivateKey | bsv.PrivateKey;
+  private xPubString?: string;
+  private url: string;
```
Comment (dorzepowski) [inline src/requester.ts:12]: Maybe `baseUrl` or `serverUrl`?

File: src/requester.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+  static CreateSigningRequester(
```
Comment (dorzepowski) [inline src/requester.ts:14]: Can it be lowercase?

File: src/requester.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    signingKey?: bsv.HDPrivateKey | bsv.PrivateKey,
+    adminKey?: bsv.HDPrivateKey,
+  ) {
+    const requester = new Requester(logger, url);
+    requester.signingKey = signingKey;
+    requester.adminKey = adminKey;
+    return requester;
+  }
+
+  static CreateXPubRequester(logger: Logger, url: string, xPubString: string) {
```
Comment (dorzepowski) [inline src/requester.ts:26]: Can it be lowercase?

File: src/requester.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  async Request(path: string, method: string = 'GET', payload: any = null): Promise<any> {
+    this.logger.debug('Making Request', path);
+    return this.makeRequest(path, method, payload, this.signingKey);
+  }
+
+  private async makeRequest(
+    path: string,
+    method: string,
```
Comment (dorzepowski) [inline src/requester.ts:52]: IMHO method here and above should be typed string like:
```
type HttpMethod = 'GET' | 'POST' ...
```

File: src/requester.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    requester.xPubString = xPubString;
+    return requester;
+  }
+
+  private constructor(logger: Logger, url: string) {
+    this.logger = logger;
+    this.url = url.endsWith('/') ? url : url + '/'; //make sure the url ends with a '/'
+  }
+
+  async AdminRequest(path: string, method: string = 'GET', payload: any = null): Promise<any> {
```
Comment (dorzepowski) [inline src/requester.ts:37]: Can it be lowercase?

Comment (dorzepowski) [inline src/requester.ts:37]: ### minor

I'm thinking if it wouldn't be more readable to have something like:
```
async adminGet(...)
async adminPost(...)
```
or even 🙈 to have at the caller side:
```
this.http.admin.get(...);
this.http.get(...);
```

File: src/requester.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    if (!this.adminKey) {
+      throw new ErrorNoAdminKey();
+    }
+    this.logger.debug('Making AdminRequest', path);
```
Comment (dorzepowski) [inline src/requester.ts:41]:
```suggestion
    this.logger.debug(`Making request as admin on ${method} ${path}`);
```

**Review by dorzepowski — CHANGES_REQUESTED** [review]

File: src/types.ts [hunk 11]
```diff
@@ -528,73 +528,99 @@ export interface AdminStats {
...
+      /**
+       * Making request with the 'xPub' doesn't sign requests, which means that the xPub will be placed into the request header.
+       */
+      signRequest: false;
+
+      /**
+       * With xPub you can make non-admin requests without signing.
+       * The SendToRecipients function will not work because it requires the xPriv.
+       */
+      xPub: string;
```
Comment (dorzepowski) [inline src/types.ts:598]: As in the code, actually if we have xPriv, then we're still able to get xPub and therefore such an options should also be valid

**Review by dorzepowski — CHANGES_REQUESTED** [review]

File: src/client.ts [hunk 12]
```diff
@@ -774,9 +736,9 @@ export class SpvWalletClient {
    * const tx = await spvWalletClient.RecordTransaction(finalized, draft.id, metadata)
    */
   async SendToRecipients(recipients: Recipients, metadata: Metadata): Promise<Transaction> {
-    const draft = await this.DraftToRecipients(recipients, metadata)
-    const finalized = this.FinalizeTransaction(draft)
-    return this.RecordTransaction(finalized, draft.id, metadata)
+    const draft = await this.DraftToRecipients(recipients, metadata);
+    const finalized = this.FinalizeTransaction(draft);
```
Comment (arkadiuszos4chain) [inline src/client.ts:740]: IMO name `FinalizeTransaction` is misleading, maybe we could change it to `SignTransaction`? WDY?

File: src/client.ts (hunk 1 above)
Comment (chris-4chain) [reply src/client.ts:755]: I've removed this line and used the `bsv.Transaction.Input` inline

File: src/client.ts (hunk 2 above)
Comment (chris-4chain) [reply src/client.ts:69]: New `ClientOptions` has been applied - as we discussed 😄

File: src/client.ts (hunk 3 above)
Comment (chris-4chain) [reply src/client.ts:55]: In current version it is no longer there.

File: src/requester.ts (hunk 9 above)
Comment (arkadiuszos4chain) [inline src/requester.ts:37]: Maybe we could introduce two client types: "normal" and "admin", which would extand the normal client.
Thanks to this in requester we would have only one "request" method, and checking adminKey would be performed on client creation

File: src/requester.ts (hunk 4 above)
Comment (chris-4chain) [reply src/requester.ts:6]: So, `HttpClient` it is 👍

File: src/requester.ts (hunk 6 above)
Comment (chris-4chain) [reply src/requester.ts:14]: I've completely removed the static methods.

File: src/requester.ts (hunk 7 above)
Comment (chris-4chain) [reply src/requester.ts:26]: I've completely removed the static methods.

File: src/requester.ts (hunk 8 above)
Comment (chris-4chain) [reply src/requester.ts:52]: Defined `type HttpMethod`

File: src/client.ts (hunk 12 above)
Comment (chris-4chain) [reply src/client.ts:740]: It's not my methods 🙈 
However I agree with you, so `SignTransaction` let it be 😄

File: src/types.ts (hunk 11 above)
Comment (chris-4chain) [reply src/types.ts:598]: New `ClientOptions` has been applied - as we discussed 😄

File: src/requester.ts (hunk 9 above)
Comment (chris-4chain) [reply src/requester.ts:37]: I've been talking to @dorzepowski about that. As I remember we decided not to do that. 
IMHO we can discuss it once again and maybe do it in some next task.

Generally, I like the idea 😄

File: src/client.ts [hunk 13]
```diff
@@ -22,112 +21,65 @@ import {
...
-      options.xPriv = undefined
-      options.xPrivString = undefined
-      options.xPub = undefined
-      options.xPubString = undefined
-      options.accessKeyString = options.accessKey.toString()
-    }
+  private xPriv?: bsv.HDPrivateKey;
 
-    this.options = options
+  constructor(serverUrl: string, options: ClientOptions, loggerConfig: LoggerConfig = defaultLogger) {
```
Comment (dorzepowski) [inline src/client.ts:50]: final touch, to have this API resilient

```suggestion
  constructor(serverUrl: string, options: XpubWithoutSigning, loggerConfig?: LoggerConfig);
  constructor(serverUrl: string, options: AccessKeyWithSigning, loggerConfig?: LoggerConfig);
  constructor(serverUrl: string, options: XprivWithSigning, loggerConfig?: LoggerConfig);
  constructor(serverUrl: string, options: AdminKey, loggerConfig?: LoggerConfig);
  constructor(serverUrl: string, options: ClientOptions, loggerConfig: LoggerConfig = defaultLogger) {
```

**Review by dorzepowski — CHANGES_REQUESTED** [review]

File: src/client.ts (hunk 12 above)
Comment (dorzepowski) [reply src/client.ts]: We should remember about this change and apply it later on go-client

File: src/requester.ts (hunk 9 above)
Comment (dorzepowski) [reply src/requester.ts]: What we agreed was that, we can do something like that, but rather in a separate task, to finish this part to enable other works on the client, like bsv.js replacement.

File: src/client.ts (hunk 13 above)
Comment (chris-4chain) [reply src/client.ts:58]: Done

**Review by dorzepowski — APPROVED** [review]

**Review by arkadiuszos4chain — APPROVED** [review]

**Comment by Nazarii-4chain** [comment]
Before it show this message for now results:
`Status: 500, Message: "no results found"`
Now it shows:
Using XPub. SendToRecipients function will not be available.
/Users/john/newSpvWalletJsClient/node_modules/@bsv/spv-wallet-js-client/dist/typescript-npm-package.cjs.js:394
                throw new ErrorResponse(this.logger, res, rawContent);
                      ^

ErrorResponse: Received error response
    at HttpClient.<anonymous> (/Users/john/newSpvWalletJsClient/node_modules/@bsv/spv-wallet-js-client/dist/typescript-npm-package.cjs.js:394:23)
    at Generator.next (<anonymous>)
    at fulfilled (/Users/john/newSpvWalletJsClient/node_modules/@bsv/spv-wallet-js-client/dist/typescript-npm-package.cjs.js:183:58)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  response: Response {
    [Symbol(realm)]: null,
    [Symbol(state)]: {
      aborted: false,
      rangeRequested: false,
      timingAllowPassed: true,
      requestIncludesCredentials: true,
      type: 'default',
      status: 500,
      timingInfo: {
        startTime: 671.509834,
        redirectStartTime: 0,
        redirectEndTime: 0,
        postRedirectStartTime: 671.509834,
        finalServiceWorkerStartTime: 0,
        finalNetworkResponseStartTime: 0,
        finalNetworkRequestStartTime: 0,
        endTime: 915.032755,
        encodedBodySize: 0,
        decodedBodySize: 0,
        finalConnectionTimingInfo: null
      },
      cacheState: '',
      statusText: 'Internal Server Error',
      headersList: HeadersList {
        cookies: null,
        [Symbol(headers map)]: Map(7) {
          'access-control-allow-credentials' => { name: 'Access-Control-Allow-Credentials', value: 'true' },
          'access-control-allow-headers' => {
            name: 'Access-Control-Allow-Headers',
            value: 'Content-Type,Cache-Control,x-auth-xpub,x-auth-key,x-auth-signature,x-auth-hash,x-auth-nonce,x-auth-time'
          },
          'access-control-allow-methods' => {
            name: 'Access-Control-Allow-Methods',
            value: 'GET, POST, PUT, DELETE, OPTIONS'
          },
          'access-control-allow-origin' => { name: 'Access-Control-Allow-Origin', value: '*' },
          'content-type' => {
            name: 'Content-Type',
            value: 'application/json; charset=utf-8'
          },
          'date' => { name: 'Date', value: 'Mon, 08 Apr 2024 07:51:15 GMT' },
          'content-length' => { name: 'Content-Length', value: '18' }
        },
        [Symbol(headers map sorted)]: null
      },
      urlList: [ URL {} ],
      body: {
        stream: ReadableStream {
          [Symbol(kType)]: 'ReadableStream',
          [Symbol(kState)]: {
            disturbed: true,
            reader: [ReadableStreamDefaultReader],
            state: 'closed',
            storedError: undefined,
            stream: undefined,
            transfer: [Object],
            controller: [ReadableStreamDefaultController]
          },
          [Symbol(nodejs.webstream.isClosedPromise)]: {
            promise: [Promise],
            resolve: [Function (anonymous)],
            reject: [Function (anonymous)]
          },
          [Symbol(nodejs.webstream.controllerErrorFunction)]: [Function: bound error]
        }
      }
    },
    [Symbol(headers)]: Headers {
      [Symbol(headers list)]: HeadersList {
        cookies: null,
        [Symbol(headers map)]: Map(7) {
          'access-control-allow-credentials' => { name: 'Access-Control-Allow-Credentials', value: 'true' },
          'access-control-allow-headers' => {
            name: 'Access-Control-Allow-Headers',
            value: 'Content-Type,Cache-Control,x-auth-xpub,x-auth-key,x-auth-signature,x-auth-hash,x-auth-nonce,x-auth-time'
          },
          'access-control-allow-methods' => {
            name: 'Access-Control-Allow-Methods',
            value: 'GET, POST, PUT, DELETE, OPTIONS'
          },
          'access-control-allow-origin' => { name: 'Access-Control-Allow-Origin', value: '*' },
          'content-type' => {
            name: 'Content-Type',
            value: 'application/json; charset=utf-8'
          },
          'date' => { name: 'Date', value: 'Mon, 08 Apr 2024 07:51:15 GMT' },
          'content-length' => { name: 'Content-Length', value: '18' }
        },
        [Symbol(headers map sorted)]: null
      },
      [Symbol(guard)]: 'immutable',
      [Symbol(realm)]: null
    }
  },
  content: '"no results found"'
}

----

Error with xpubString

const client = new SpvWalletClient(server, {
    signRequest: true,
    xPubString:"xpub..."
})
Is it possible to use xPubString ?
   ErrorInvalidOptions: Invalid options. None of xPub, xPriv, accessKey nor adminKey is set
   
   ---
    Not sure if this related to your changes
   xPriv Error
   const keys = generateKeys()
   
   const client = new SpvWalletClient(server, {
     xPriv: "string"
})
  Using xPriv to sign requests
/Users/john/newSpvWalletJsClient/node_modules/@bsv/spv-wallet-js-client/dist/typescript-npm-package.cjs.js:428
            this.xPriv = bsv.HDPrivateKey.fromString(options.xPriv);
                                          ^

TypeError: Cannot read properties of undefined (reading 'fromString')
    at SpvWalletClient.makeRequester (/Users/john/newSpvWalletJsClient/node_modules/@bsv/spv-wallet-js-client/dist/typescript-npm-package.cjs.js:428:43)

**Pull request merged**